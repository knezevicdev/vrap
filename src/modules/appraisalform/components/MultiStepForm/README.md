### MultiStepForm Component

The purpose of this component is to create a wizard-like experience for our forms, breaking them down into sections.

### Required Props

1. formTitle - String; Title for your Multi Step Form ex: 'Multi-step form title'
2. sections - Collection; This is the meat of the props passed in, this is an array of each section you'll be passing to the form.
   Each section should contain a component, form, subTitle, and title.
   Component is the actual component you'll be rendered for that step.
   Form is the form supplied by the useForm custom hook
   SubTitle to title the form
   Title is to be a title for the whole section
   ex: {
   component: FirstInfoForm,
   form: firstForm,
   subTitle: 'Form 1',
   title: 'Section 1'
   }
3. onDone - Function that is executed once the final section is completed. Here is where any redux or API calls will take place
4. active - Integer that dictates which section is shown first; defaults to 0

### Example usage of MultiStepForm

<MultiStepForm
  formTitle="Multi-step form example"
  sections={sections}
  onDone={onSubmit}
  active={0}
/>

### Form Components

In the adjacent example.js file you'll see very simple Forms

ex:
const FirstInfoForm = ({ fields }) => (
<>
<FirstNameInput field={fields.first} />
<LastNameInput field={fields.last} />
<EmailInput field={fields.email} />
</>
);

The idea here is the actual forms, which will live as their own independent components, will do nothing but contain the HTML
for that section.
The Component piecing together the form will handle redux, state, and APIs (see FormMain in example.js)

### Unit Tests

Today we only have a snapshot test for this component since Enzyme/Jest/React doesn't yet support hooks, so once that happens
we'll have to come back to this component.
The tests we really _WANT_ for this component is the logic within the `handleOnNext` function.
Within `handleOnNext` we will want to test that the button updates to "Review" when the final section is active. We'll also
want to test that when clicking the button while on the final step we call `onDone`, and when we aren't on the final step we
just increment the activeSection state value.

### To view example.js

url: /multiformComponent
