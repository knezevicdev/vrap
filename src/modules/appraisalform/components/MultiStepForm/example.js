import React from 'react';
import useForm from '@app/components/Form/useForm';
import FirstNameInput from '@app/components/Form/Inputs/FirstNameInput';
import LastNameInput from '@app/components/Form/Inputs/LastNameInput';
import EmailInput from '@app/components/Form/Inputs/EmailInput';
import MultiStepForm from './';
import PropTypes from 'prop-types';

export const FirstInfoForm = ({ fields }) => (
  <>
    <FirstNameInput field={fields.first} />
    <LastNameInput field={fields.last} />
    <EmailInput field={fields.email} />
  </>
);

export const SecondInfoForm = ({ fields }) => (
  <>
    <FirstNameInput field={fields.first} />
    <LastNameInput field={fields.last} />
    <EmailInput field={fields.email} />
  </>
);

export const ThirdInfoForm = ({ fields }) => (
  <>
    <FirstNameInput field={fields.first} />
    <LastNameInput field={fields.last} />
    <EmailInput field={fields.email} />
  </>
);

FirstInfoForm.propTypes = {
  fields: PropTypes.object
};

SecondInfoForm.propTypes = {
  fields: PropTypes.object
};

ThirdInfoForm.propTypes = {
  fields: PropTypes.object
};

const FormMain = () => {
  const sendToRedux = () => {
    // console.log('reduxed');
  };

  const defaultFieldValues = {
    defaultValues: {
      first: '',
      last: '',
      email: ''
    },
    customEvents: {
      onBlur: sendToRedux
    }
  };

  const firstForm = useForm(defaultFieldValues);
  const secondForm = useForm(defaultFieldValues);
  const thirdForm = useForm(defaultFieldValues);

  const sections = [
    {
      component: FirstInfoForm,
      form: firstForm,
      subTitle: 'Form 1',
      title: 'Section 1'
    },
    {
      component: SecondInfoForm,
      form: secondForm,
      subTitle: 'Form 2',
      title: 'Section 2'
    },
    {
      component: ThirdInfoForm,
      form: thirdForm,
      subTitle: 'Form 3',
      title: 'Section 3'
    }
  ];

  const onSubmit = () => {
    // get field values for each "form"
    //     const firstFormFields = firstForm.fields;
    //     const secondFormFields = secondForm.fields;
    //     const thirdFormFields = thirdForm.fields;

    alert("Here is where we'd submit to api and add to redux if necessary");
  };

  return (
    <MultiStepForm
      formTitle="Multi-step form example"
      sections={sections}
      onDone={onSubmit}
      active={0}
    />
  );
};

export default FormMain;
