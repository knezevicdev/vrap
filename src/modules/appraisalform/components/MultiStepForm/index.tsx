import { default as MultiStepForm } from './MultiStepForm';
export default MultiStepForm;
