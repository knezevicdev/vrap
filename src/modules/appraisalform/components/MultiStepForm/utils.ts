export const greenCheckPath =
  'https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/check-mark-green.svg';
const one_icon = `https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/1-active.svg`;
const two_icon = `https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/2-active.svg`;
const three_icon = `https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/3-active.svg`;
const four_icon = `https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/4-active.svg`;
const five_icon = `https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/5-active.svg`;
const six_icon = `https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/6-active.svg`;
const gray_two_icon = `https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/2-inactive.svg`;
const gray_three_icon = `https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/3-inactive.svg`;
const gray_four_icon = `https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/4-inactive.svg`;
const gray_five_icon = `https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/5-inactive.svg`;
const gray_six_icon = `https://assets.vroomcdn.com/static-rebrand/icons/svg/400-gt/6-inactive.svg`;

export const blueIcons: any = {
  0: one_icon,
  1: two_icon,
  2: three_icon,
  3: four_icon,
  4: five_icon,
  5: six_icon,
};

export const grayIcons: any = {
  1: gray_two_icon,
  2: gray_three_icon,
  3: gray_four_icon,
  4: gray_five_icon,
  5: gray_six_icon,
};
