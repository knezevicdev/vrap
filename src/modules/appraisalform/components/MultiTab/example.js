import React from 'react';
import PropTypes from 'prop-types';
import MultiTab from './';

export const FirstInfoTab = () => (
  <>
    <span>tab 1</span>
  </>
);

export const SecondInfoTab = ({ props }) => (
  <>
    <span>tab 2</span>
    {props.text}
  </>
);

const TabMain = () => {
  const tabSections = [
    {
      component: FirstInfoTab,
      title: 'tab 1'
    },
    {
      component: SecondInfoTab,
      props: { text: 'passed in custom prop' },
      title: 'tab 2'
    }
  ];

  return <MultiTab tabSections={tabSections} active={0} />;
};

SecondInfoTab.propTypes = {
  props: PropTypes.object,
  text: PropTypes.string
};

export default TabMain;
