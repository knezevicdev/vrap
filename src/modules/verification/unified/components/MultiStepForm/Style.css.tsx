import { addStyleForMobile, Button } from '@vroom-web/ui-lib';
import React from 'react';
import styled from 'styled-components';

export const FormHeader = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const Block = styled.div`
  display: none;
  @media (max-width: 420px) {
    display: block;
  }
`;

export const FormTitle = styled.h1`
  padding: 20px 0;
  font-family: Vroom-Sans, sans-serif;
  font-size: 36px;
  line-height: 40px;
  letter-spacing: 1px;
  text-align: left;
  margin: 0 0 8px 0;

  ${addStyleForMobile(`
      text-align: center;
      padding-bottom: 16px;
      padding-top: 15px;
      font-family: Vroom-Sans;
      font-size: 28px;
      line-height: 28px;
      letter-spacing: 1px;
    `)}
`;

export const FormStep = styled.div`
  padding: 20px 0;
  border-top: 1px solid #e0e0e0;
`;

export const FormSection = styled(({ ...restProps }) => <div {...restProps} />)`
  padding-top: 0;
  transition: max-height 300ms ease-in-out;
  height: auto;
  max-height: 0;
  overflow: ${(props) => (props.isactive === 'true' ? 'initial' : 'hidden')};

  > div {
    transition: max-height 300ms ease-in-out;
    max-height: 0;
  }

  /* big max-height is used when animating variable height transitions */
  ${(props) =>
    props.isactive === 'true' &&
    `
      max-height: 4500px;

      > div {
        max-height: 4500px;
      }
    `}
`;

export const SectionHeader = styled.div`
  display: flex;
  flex-wrap: wrap;
  padding-bottom: 20px;
`;

export const StepNumber = styled.div`
  padding-right: 10px;
  margin: auto 0;

  img {
    vertical-align: middle;
    border-style: none;
  }
`;

export const StepTitle = styled(({ ...restProps }) => <div {...restProps} />)`
  font-family: Calibre-Semibold, sans-serif;
  font-size: 20px;
  line-height: 26px;
  letter-spacing: 0.25px;
  color: #041022;

  ${(props) => props.isactive === 'false' && `color: #999da3;`}
  ${addStyleForMobile(`
    font-size: 18px;
  `)}
`;

export const EditStep = styled.div`
  padding-left: 10px;
  color: #e7131a;
  cursor: pointer;
  line-height: 26px;
  font-family: Calibre-Regular, sans-serif;
  font-size: 16px;
`;

export const NextButton = styled(({ ...restProps }) => (
  <Button.Primary {...restProps} />
))`
  :enabled {
    background-color: ${(props) =>
      props.isCTAColorExp ? '#308406' : '#E7131A'};
    &:hover {
      background-color: ${(props) =>
        props.isCTAColorExp ? '#309706' : '#d01118'};
    }
  }
  width: auto;
  min-width: 180px;
  margin: 20px 0;

  ${addStyleForMobile(`
    width: 100%;
  `)};
`;
