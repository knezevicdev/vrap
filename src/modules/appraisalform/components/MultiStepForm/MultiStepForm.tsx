import { addStyleForMobile, Tooltip } from '@vroom-web/ui-lib';
import { Button } from '@vroom-web/ui-lib';
import React, { useEffect, useState } from 'react';
import styled from 'styled-components';

import { GenericObject } from '../../../../interfaces.d';
import PrevNextButtons from '../PrevNextButtons';
import { blueIcons, grayIcons, greenCheckPath } from './utils';

function numberIcon(index: number, activeSection: number, className: string) {
  const isActive = activeSection === index;
  let src = '';
  if (isActive) {
    src = blueIcons[index];
  } else {
    src = grayIcons[index];
  }
  return <img alt="" src={src} className={className} />;
}

interface Props {
  formTitle: string;
  formSubtittleSection?: () => void;
  sections: any[];
  onDone: () => void | Promise<void>;
  onNext: (
    currentSection: number,
    nextSection: number,
    clear?: boolean
  ) => void;
  onNextIntercept?: () => void;
  active: number;
  customAnalyticsFunc?: (
    nextTitle: string | undefined,
    currentTitle: string
  ) => void;
  refreshButton?: boolean;
  showDialog?: () => void;
  nextText: string;
  submitText: string;
  appraisalTitle?: string;
  disableExperiments: boolean;
  showSteps?: boolean;
}

const MultiStepForm: React.FC<Props> = (props) => {
  const {
    formTitle,
    formSubtittleSection,
    sections,
    onDone,
    onNext,
    active = 0,
    customAnalyticsFunc,
    refreshButton = false,
    nextText = 'Next',
    submitText = 'Review',
    appraisalTitle,
    disableExperiments,
    showSteps = false,
  } = props;

  useEffect(() => {
    window.scrollTo(0, 0);
  }, []);

  const maxSteps = sections.length - 1;
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [activeSection, setActiveSection] = useState(active);
  const [returnSection, setReturnSection] = useState(null);
  const [hideButton, setHideButton] = useState<boolean>(false);

  // using a static string here to swap values but this can easily be passed in props
  // this can also be handled by parent component if we don't like this
  const [buttonText, setButtonText] = useState(nextText);

  useEffect(() => {
    // when returning to the form via review page we usually like to update the buttons
    // if next and submit are equal after changing we update button state
    // this implementation is specific to review/edit scenarios and is welcome for
    // a better solution if someone has one
    if (nextText == submitText) {
      setButtonText(submitText);
    }
  }, [nextText, submitText]);

  useEffect(() => {
    if (location.hash !== '') {
      setButtonText(submitText);
    } else {
      if (activeSection === sections.length - 1) {
        setButtonText(submitText);
      } else {
        setButtonText(nextText);
      }
    }
  }, [activeSection]);

  useEffect(() => {
    // SetTimeout is used here to make sure we're waiting to
    // scrollTo ONLY after the height animation finishes
    // then and only then we scroll to the top of the next active section
    // when active section isn't 0
    if (activeSection !== 0) {
      setTimeout(() => {
        const activeElement: any = document.getElementById(
          activeSection as any
        );
        const headerOffset = 70;

        if (activeElement) {
          const elementPosition = activeElement.offsetTop;
          const offsetPosition = elementPosition - headerOffset;

          window.scrollTo({
            top: offsetPosition,
            behavior: 'smooth',
          });
        }
      }, 500);
    }
  }, [activeSection]);

  const handleRefreshButton = () => {
    sections.forEach((section: GenericObject) => {
      section.form.resetForm();
    });

    setActiveSection(0);
    setButtonText(nextText);
    onNext(activeSection, 0, true);
  };

  const handleRefreshKeyDown = (e: any) => {
    if (e.key === 'Enter') {
      sections.forEach((section: GenericObject) => {
        section.form.resetForm();
      });

      setActiveSection(0);
      setButtonText(nextText);
      onNext(activeSection, 0, true);
    }
  };

  const handleOnNext = (_e: any, onNextIntercept: any) => {
    const nextStep = activeSection + 1;
    let currentSectionIsValid = sections[activeSection].form.isFormValid;
    //const nextSection = sections[nextStep];
    if (sections[activeSection].addressForm) {
      currentSectionIsValid =
        sections[activeSection].addressForm.isFormValid &&
        sections[activeSection].form.isFormValid;
    }
    if (customAnalyticsFunc) {
      customAnalyticsFunc(
        sections[activeSection + 1]
          ? sections[activeSection + 1].title
          : undefined,
        sections[activeSection].title
      );
    }

    if (onNextIntercept) {
      onNextIntercept(handleOnNext);
      return;
    }

    setIsSubmitting(true);

    if (nextStep > maxSteps && currentSectionIsValid) {
      onDone();
      setIsSubmitting(false);
    } else if (returnSection) {
      onNext && onNext(activeSection, returnSection);
      setActiveSection(returnSection);
      setReturnSection(null);
      setIsSubmitting(false);
    } else {
      onNext && onNext(activeSection, nextStep);
      setActiveSection(nextStep);
      setIsSubmitting(false);
    }
  };

  const handleEditSection = (sectionToOpen: any, newReturnSection: any) => {
    //const currentSection = sections[sectionToOpen];
    setButtonText(submitText);
    setActiveSection(sectionToOpen);
    setReturnSection(newReturnSection);
  };

  const handleEditKeyDown = (
    event: any,
    sectionToOpen: any,
    newReturnSection: any
  ) => {
    if (event.key === 'Enter') {
      setButtonText(submitText);
      setActiveSection(sectionToOpen);
      setReturnSection(newReturnSection);
    }
  };

  const getNextButton = (
    formComponent: any,
    formIsInvalid: any,
    onNextIntercept: any
  ) => {
    if (
      formComponent.toolTip &&
      formIsInvalid &&
      formComponent.toolTip.showTooltip
    ) {
      return (
        <>
          <Tooltip content={<span>{formComponent.toolTip.content}</span>} />
          <NextButtonWrapper>
            <NextButton
              onClick={handleOnNext}
              disabled={formIsInvalid}
              data-qa={'Continue'}
            >
              {buttonText}
            </NextButton>
          </NextButtonWrapper>
        </>
      );
    } else {
      return (
        <NextButton
          onClick={(e: GenericObject) => handleOnNext(e, onNextIntercept)}
          disabled={formIsInvalid}
          data-qa={'Continue'}
        >
          {buttonText}
        </NextButton>
      );
    }
  };

  const getPrevNextButtons = (formComponent: any, formIsInvalid: any) => {
    const { prevStep } = formComponent;
    if (
      formComponent.toolTip &&
      formIsInvalid &&
      formComponent.toolTip.showTooltip
    ) {
      return (
        <>
          <Tooltip content={<span>{formComponent.toolTip.content}</span>} />
          <NextButtonWrapper>
            <PrevNextButtons
              goToPrevStep={prevStep}
              goToNextStep={handleOnNext}
              nextDisabled={formIsInvalid}
              nextButtonLabel={buttonText}
            />
          </NextButtonWrapper>
        </>
      );
    } else {
      return (
        <PrevNextButtons
          goToPrevStep={prevStep}
          goToNextStep={handleOnNext}
          nextDisabled={formIsInvalid}
          nextButtonLabel={buttonText}
        />
      );
    }
  };

  const hideButtonCallback = (hide: boolean): void => {
    setHideButton(hide);
  };

  const formComponents = sections.map((formComponent: any, idx: any) => {
    const {
      component,
      form: { fields },
      prevStep,
      showDialog,
      title,
      data,
      onNextIntercept,
    } = formComponent;
    const CurrentComponent = component;
    const isActive = idx === activeSection;
    let formIsInvalid =
      !sections[activeSection].form.isFormValid || isSubmitting;
    if (sections[activeSection].addressForm) {
      formIsInvalid =
        !sections[activeSection].form.isFormValid ||
        isSubmitting ||
        !sections[activeSection].addressForm.isFormValid;
    }
    const showGreenCheck = idx < activeSection;
    const showEditButton = !isActive && idx < activeSection;

    return (
      <FormStep key={idx} id={idx}>
        <SectionHeader>
          <StepNumber>
            {showGreenCheck && (
              <img alt="" src={greenCheckPath} className={'title-icon'} />
            )}
            {!showGreenCheck && numberIcon(idx, activeSection, 'title-icon')}
          </StepNumber>
          <StepTitle isactive={isActive.toString()}>{title}</StepTitle>
          {showEditButton && (
            <EditStep
              role="button"
              tabIndex={0}
              onClick={() => handleEditSection(idx, activeSection)}
              onKeyDown={(event) =>
                handleEditKeyDown(event, idx, activeSection)
              }
            >
              Edit
            </EditStep>
          )}
          {!showEditButton && showSteps && (
            <TimeEst>
              (Step {idx + 1} of {sections.length})
            </TimeEst>
          )}
        </SectionHeader>
        <FormSection isactive={isActive.toString()}>
          {formComponent.subTitle && <h3>{formComponent.subTitle}</h3>}
          <CurrentComponent
            fields={fields}
            showDialog={showDialog}
            form={formComponent.form}
            data={data}
            {...formComponent}
            disableExperiments={disableExperiments}
            hideButtonCallback={hideButtonCallback}
          />
          {!hideButton &&
            (prevStep
              ? getPrevNextButtons(formComponent, formIsInvalid)
              : getNextButton(formComponent, formIsInvalid, onNextIntercept))}
        </FormSection>
      </FormStep>
    );
  });

  return (
    <div>
      <FormHeader>
        <Block />
        <FormTitle>
          <span>{formTitle}</span>
          {formSubtittleSection && formSubtittleSection()}
        </FormTitle>
        {refreshButton && (
          <FormRefresh
            role="button"
            tabIndex={0}
            onClick={handleRefreshButton}
            onKeyDown={handleRefreshKeyDown}
          >
            Refresh
          </FormRefresh>
        )}
      </FormHeader>
      {appraisalTitle && <FormWarning>{appraisalTitle}</FormWarning>}
      {formComponents}
    </div>
  );
};

const FormHeader = styled.div`
  display: flex;
  justify-content: space-between;
`;

const Block = styled.div`
  display: none;
  @media (max-width: 420px) {
    display: block;
  }
`;

const FormTitle = styled.h1`
  padding: 20px 0px;
  font-family: Vroom-Sans;
  font-size: 36px;
  line-height: 40px;
  letter-spacing: 1px;
  text-align: left;
  margin: 0 0 8px 0;

  ${addStyleForMobile(`
      text-align: center;
      padding-bottom: 16px;
      padding-top: 15px;
      font-family: Vroom-Sans;
      font-size: 28px;
      line-height: 28px;
      letter-spacing: 1px;
    `)}
`;

const FormRefresh = styled.div`
  margin: auto 0;
  color: #e7131a;
  font-weight: bold;
  cursor: pointer;
  font-family: Calibre-Regular;
`;

const FormStep = styled.div`
  padding: 20px 0px;
  border-top: 1px solid #e0e0e0;
`;

const FormSection = styled(({ ...restProps }) => <div {...restProps} />)`
  padding-top: 0;
  transition: max-height 300ms ease-in-out;
  height: auto;
  max-height: 0;
  overflow: ${(props) => (props.isactive === 'true' ? 'initial' : 'hidden')};

  > div {
    transition: max-height 300ms ease-in-out;
    max-height: 0;
  }

  /* big max-height is used when animating variable height transitions */
  ${(props) =>
    props.isactive === 'true' &&
    `
      max-height: 4500px;

      > div {
        max-height: 4500px;
      }
    `}
`;

const SectionHeader = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

const StepNumber = styled.div`
  padding-right: 10px;
  margin: auto 0;

  img {
    vertical-align: middle;
    border-style: none;
  }
`;

const StepTitle = styled(({ ...restProps }) => <div {...restProps} />)`
  font-family: Calibre-Semibold;
  font-size: 20px;
  line-height: 26px;
  letter-spacing: 0.25px;
  color: #041022;

  ${(props) => props.isactive === 'false' && `color: #999da3;`}
  ${addStyleForMobile(`
    font-size: 18px;
  `)}
`;

const EditStep = styled.div`
  padding-left: 10px;
  color: #e7131a;
  cursor: pointer;
  line-height: 26px;
  font-family: Calibre-Regular;
  font-size: 16px;
`;

const TimeEst = styled.div`
  font-family: Calibre-Regular;
  font-size: 18px;
  letter-spacing: 0.25px;
  line-height: 25px;
  color: #999da3;
  padding-left: 5px;
  white-space: nowrap;
`;

const NextButton = styled(({ ...restProps }) => (
  <Button.Primary {...restProps} />
))`
  :enabled {
    background-color: #e7131a;

    &:hover {
      background-color: #d01118;
    }
  }
  width: auto;
  min-width: 180px;
  margin: 20px 0;

  ${addStyleForMobile(`
    width: 100%;
  `)};
`;

const NextButtonWrapper = styled.span`
  padding: 10px;
`;

const FormWarning = styled.div`
  font-size: 18px;
  font-family: Calibre-Regular;
  line-height: 24px;
  display: flex;
  align-items: center;
  letter-spacing: 0.25px;
  margin-bottom: 24px;
  text-align: left;
  @media (max-width: 420px) {
    text-align: center;
  }
`;

export default MultiStepForm;
