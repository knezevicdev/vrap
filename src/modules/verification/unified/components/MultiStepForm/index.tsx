import React, { useEffect, useRef, useState } from 'react';

import {
  Block,
  EditStep,
  FormHeader,
  FormSection,
  FormStep,
  FormTitle,
  NextButton,
  SectionHeader,
  StepNumber,
  StepTitle,
} from './Style.css';
import { blueIcons, grayIcons, greenCheckPath } from './utils';

function numberIcon(index: number, activeSection: number, className: string) {
  const isActive = activeSection === index;
  const src = isActive ? blueIcons[index] : grayIcons[index];
  return <img alt="" src={src} className={className} />;
}

export interface FormStepProps {
  nextStep: () => void;
  goToStep: (step: number) => void;
  editRef?: number;
}

export interface FormSection {
  component: React.ComponentType<FormStepProps>;
  isValid: boolean;
  title: string;
  disableNext?: boolean;
  completedAfterComponent?: React.ComponentType;
}

interface Props {
  formTitle: string;
  sections: FormSection[];
  onDone: () => void | Promise<void>;
  onNext: (currentSection: number, nextSection: number) => void | Promise<void>;
  onSectionChange?: (section: number) => void;
  active: number;
  nextText: string;
  submitText: string;
}

const MultiStepForm: React.FC<Props> = (props) => {
  const {
    formTitle,
    sections,
    onDone,
    onNext,
    onSectionChange,
    active = 0,
    nextText = 'Next',
  } = props;

  useEffect(() => {
    window.scrollTo(0, 0);
  }, []);

  const maxSteps = sections.length - 1;
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [activeSection, setActiveSection] = useState(active);
  const [returnSection, setReturnSection] = useState(0);
  const editRefs = useRef<Record<number, number>>({});

  useEffect(() => {
    if (activeSection !== 0) {
      setTimeout(() => {
        const activeElement = document.getElementById(String(activeSection));
        const headerOffset = 70;

        if (activeElement) {
          const elementPosition = activeElement.offsetTop;
          const offsetPosition = elementPosition - headerOffset;

          window.scrollTo({
            top: offsetPosition,
            behavior: 'smooth',
          });
        }
      }, 500);
    }
  }, [activeSection]);

  const handleOnNext = () => {
    const nextStep = returnSection ? returnSection : activeSection + 1;
    const isCurrentSectionValid = sections[activeSection].isValid;
    if (!isCurrentSectionValid) return;
    if (nextStep > maxSteps) {
      onDone();
      return;
    }

    setIsSubmitting(true);

    onNext?.(activeSection, nextStep);
    setActiveSection(nextStep);
    onSectionChange?.(nextStep);
    if (returnSection) setReturnSection(0);

    setIsSubmitting(false);
  };

  const handleGoToStep = (step: number) => {
    if (activeSection === step) return;

    setIsSubmitting(true);

    setActiveSection(step);
    onSectionChange?.(step);
    if (step === returnSection) {
      setReturnSection(0);
    }

    setIsSubmitting(false);
  };

  const handleEditSection = (
    sectionToOpen: number,
    newReturnSection: number
  ) => {
    if (!editRefs.current[sectionToOpen]) editRefs.current[sectionToOpen] = 0;
    editRefs.current[sectionToOpen]++;
    setActiveSection(sectionToOpen);
    onSectionChange?.(sectionToOpen);
    setReturnSection(newReturnSection);
  };

  const handleEditKeyDown = (
    event: React.KeyboardEvent<HTMLDivElement>,
    sectionToOpen: number,
    newReturnSection: number
  ) => {
    if (event.key === 'Enter') {
      handleEditSection(sectionToOpen, newReturnSection);
    }
  };

  const formComponents = sections.map(
    (formComponent: FormSection, idx: number) => {
      const { component, title, disableNext, completedAfterComponent } =
        formComponent;
      const CurrentComponent = component;
      const AfterComponent = completedAfterComponent;
      const isActive = idx === activeSection;
      const isFormInvalid = !sections[activeSection].isValid || isSubmitting;
      const showGreenCheck = idx < activeSection;
      const showEditButton = !isActive && idx < activeSection;

      return (
        <FormStep key={idx} id={String(idx)}>
          <SectionHeader>
            <StepNumber>
              {showGreenCheck && (
                <img alt="" src={greenCheckPath} className={'title-icon'} />
              )}
              {!showGreenCheck && numberIcon(idx, activeSection, 'title-icon')}
            </StepNumber>
            <StepTitle isactive={isActive.toString()}>{title}</StepTitle>
            {showEditButton && (
              <EditStep
                role="button"
                tabIndex={0}
                onClick={() => handleEditSection(idx, activeSection)}
                onKeyDown={(event) =>
                  handleEditKeyDown(event, idx, activeSection)
                }
              >
                Edit
              </EditStep>
            )}
          </SectionHeader>
          {showEditButton && AfterComponent && (
            <FormSection isactive="true">
              <AfterComponent />
            </FormSection>
          )}
          <FormSection isactive={isActive.toString()}>
            <CurrentComponent
              nextStep={handleOnNext}
              goToStep={handleGoToStep}
              editRef={editRefs.current[idx]}
            />
            {!disableNext && (
              <NextButton
                onClick={handleOnNext}
                disabled={isFormInvalid}
                data-qa={'Continue'}
              >
                {nextText}
              </NextButton>
            )}
          </FormSection>
        </FormStep>
      );
    }
  );

  return (
    <div>
      <FormHeader>
        <Block />
        <FormTitle>
          <span>{formTitle}</span>
        </FormTitle>
      </FormHeader>
      {formComponents}
    </div>
  );
};

export default MultiStepForm;
