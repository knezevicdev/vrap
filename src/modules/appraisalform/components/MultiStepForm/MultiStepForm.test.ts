import MultiStepForm from './MultiStepForm';
import { FirstInfoForm, SecondInfoForm, ThirdInfoForm } from './example';

describe('<MultiStepForm />', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    const formExample = {
      fields: {
        first: {
          value: '',
          error: false,
          autofocus: true
        },
        last: {
          value: '',
          error: false,
          autofocus: false
        },
        email: {
          value: '',
          error: false,
          autofocus: false
        }
      },
      isFormValid: jest.fn()
    };

    const sections = [
      {
        component: FirstInfoForm,
        form: formExample,
        subTitle: 'Form 1',
        title: 'Section 1'
      },
      {
        component: SecondInfoForm,
        form: formExample,
        subTitle: 'Form 2',
        title: 'Section 2'
      },
      {
        component: ThirdInfoForm,
        form: formExample,
        subTitle: 'Form 3',
        title: 'Section 3'
      }
    ];

    props = {
      formTitle: 'Multi-step form example',
      sections: sections,
      onDone: jest.fn(),
      active: 0
    };
  });

  describe('rendering', () => {
    it('should match the snapshot', () => {
      wrapper = shallow(<MultiStepForm {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  // Ideally we'd have some tests around handleOnNext
  // however since enzyme doesn't yet support lifecycle hooks
  // this'll have to wait until a later date

  afterEach(() => {
    jest.clearAllMocks();
  });
});
