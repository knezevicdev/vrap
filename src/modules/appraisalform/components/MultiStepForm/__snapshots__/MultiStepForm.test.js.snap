// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiStepForm /> rendering should match the snapshot 1`] = `
<div>
  <styled.h1>
    Multi-step form example
  </styled.h1>
  <styled.div
    key="0"
  >
    <styled.div>
      <styled.div>
        <img
          className="title-icon"
          src="undefined/icons/svg/400-gt/1-active.svg"
        />
      </styled.div>
      <styled.div
        isActive={true}
      >
        Section 1
      </styled.div>
    </styled.div>
    <styled.div
      isActive={true}
    >
      <h3>
        Form 1
      </h3>
      <FirstInfoForm
        fields={
          Object {
            "email": Object {
              "autofocus": false,
              "error": false,
              "value": "",
            },
            "first": Object {
              "autofocus": true,
              "error": false,
              "value": "",
            },
            "last": Object {
              "autofocus": false,
              "error": false,
              "value": "",
            },
          }
        }
      />
      <Styled(Styled(Styled(BaseButtonLink)))
        disabled={false}
        onClick={[Function]}
      >
        Next
      </Styled(Styled(Styled(BaseButtonLink)))>
    </styled.div>
  </styled.div>
  <styled.div
    key="1"
  >
    <styled.div>
      <styled.div>
        <img
          className="title-icon"
          src="undefined/icons/svg/400-gt/2-inactive.svg"
        />
      </styled.div>
      <styled.div
        isActive={false}
      >
        Section 2
      </styled.div>
    </styled.div>
    <styled.div
      isActive={false}
    >
      <h3>
        Form 2
      </h3>
      <SecondInfoForm
        fields={
          Object {
            "email": Object {
              "autofocus": false,
              "error": false,
              "value": "",
            },
            "first": Object {
              "autofocus": true,
              "error": false,
              "value": "",
            },
            "last": Object {
              "autofocus": false,
              "error": false,
              "value": "",
            },
          }
        }
      />
      <Styled(Styled(Styled(BaseButtonLink)))
        disabled={false}
        onClick={[Function]}
      >
        Next
      </Styled(Styled(Styled(BaseButtonLink)))>
    </styled.div>
  </styled.div>
  <styled.div
    key="2"
  >
    <styled.div>
      <styled.div>
        <img
          className="title-icon"
          src="undefined/icons/svg/400-gt/3-inactive.svg"
        />
      </styled.div>
      <styled.div
        isActive={false}
      >
        Section 3
      </styled.div>
    </styled.div>
    <styled.div
      isActive={false}
    >
      <h3>
        Form 3
      </h3>
      <ThirdInfoForm
        fields={
          Object {
            "email": Object {
              "autofocus": false,
              "error": false,
              "value": "",
            },
            "first": Object {
              "autofocus": true,
              "error": false,
              "value": "",
            },
            "last": Object {
              "autofocus": false,
              "error": false,
              "value": "",
            },
          }
        }
      />
      <Styled(Styled(Styled(BaseButtonLink)))
        disabled={false}
        onClick={[Function]}
      >
        Next
      </Styled(Styled(Styled(BaseButtonLink)))>
    </styled.div>
  </styled.div>
</div>
`;
